File: QQMhtSpliter.py
# coding=UTF-8
#! /usr/local/bin/python3
import time
import re
import os
import sys
import base64


# dat替换为图片形式
def src(m):
    s = re.sub('.dat', '', m.group(1))
    if s in image_type_dict.keys():
      return 'src="./images/' + s + image_type_dict[s] + '"'
    else:
      dat_list.append('{}.html => {}不存在'.format(convert_name, s))
      return 'src="./images/' + s + '.gif"'
# 获取图片格式
def get_img_type(b64):
    s = b64[0:25]
    if s.find('gif') != -1:
        ext = '.gif'
    elif s.find('png') != -1:
        ext = '.png'
    elif s.find('jpg') != -1 or s.find('jpeg') != -1:
        ext = '.jpg'
    return ext
# base64转换为图片
def base64toimage(b64, k, img_type):
    c = b64.split(k + '.dat')[1].strip() #只留base64编码 其他删除
    c = base64.b64decode(c)
    with open(image_dir + k + img_type, 'wb') as f: #b 用于图片等
        f.write(c)
def read_all(dir, type):
    items_list = []
    for parent, dirnames, filenames in os.walk(dir):
        if type == 'dir' and parent != '.' and parent.find('images') == -1: #排除images文件夹
            items_list.append(parent + '/')
        if type == 'file':
            for filename in filenames:
                if filename[0] != '.' and filename.find('.mht') != -1: #只获取mht文件
                    items_list.append(filename)
    return items_list


start = time.time()
# 分割文件大小
per = input('请输入分割文件的大小（单位：M）：')
per = int(per)
# 分割符
split_str = '\n'
# 获取当前目录的mht文件列表
file_list = read_all('.', 'file')
dat_dict = {}
dat_list = []
convert_num = 0
for fl in file_list:
  convert_num = convert_num + 1
  base_dir = './convert_' + str(convert_num) + '/'
  image_dir = base_dir + 'images/'
  if not os.path.exists(image_dir):
      os.makedirs(image_dir, 0o777)
  qq_name = re.sub('.mht', '', fl)
  file_name_prefix = qq_name + '_'
  o_file = qq_name + '.mht'
  size = os.path.getsize(o_file)
  # 读取原文件
  with open(o_file, encoding="UTF-8") as f:
      qq = f.read()
  html_end = '</table></body></html>'
  # 分割文件
  content = qq.split(html_end)
  # 聊天内容(含头)
  txt_qq = content[0].strip() + html_end
  # 图片
  qq_image = content[1].strip()
  # 生成header header2
  ret = re.search('\\d{4}-\\d{2}-\\d{2}</td></tr>', txt_qq)
  split_date = ret.group()
  txt = txt_qq.split(split_date)
  header_str = txt[0] + split_date + split_str
  # 获取分割符sep
  for l in header_str.split(split_str):
      sep = l.strip()
      if sep.find('------=_NextPart_') != -1:
          break
  header_str = header_str.split('Content-Transfer-Encoding:7bit')[1].strip() + split_str
  header2_str = header_str.replace('height:24px;line-height:24px;padding-left:10px;margin-bottom:5px;', 'padding-left:10px;')
  header2_str = header2_str.replace('<tr><td><div style=padding-left:10px;>&nbsp;</div></td></tr>', '')
  header2_str = re.sub('日期: \\d{4}-\\d{2}-\\d{2}', '&nbsp;', header2_str)
  # 聊天内容
  qq_content = txt[1].replace(html_end, '').strip()
  # 图片列表
  qq_image_list = qq_image.split(sep)
  # 删除列表的第一个和最后一个
  qq_image_list.pop(0)
  qq_image_list.pop()
  image_size_dict = {}
  image_type_dict = {}
  # 生成图片文件
  for l in qq_image_list:
      dat = re.search('\\{[^\\}]+\\}\\.dat', l)
      dat_key = re.sub('\\.dat', '', dat.group())
      # 每张图片大小 单位kb
      image_size_dict[dat_key] = round(float(len(l)) / 1024, 2)
      img_type = get_img_type(l)
      image_type_dict[dat_key] = img_type
      base64toimage(l.strip(), dat_key, img_type)
  if size/1024/1024 < per:
      convert_name = base_dir + qq_name
      html = re.sub('src="(\\{[^"]+)"', src, qq_content)
      new_content = header_str + html + html_end + split_str
      with open(convert_name +'.html', 'w', encoding="UTF-8") as f:
          f.write(new_content)
      print(qq_name + ' 不需分割，将直接转换为html文件')
      continue
  else:
      # 分割聊天内容
      qq_content_list = qq_content.split(split_str)
      # 图片大小统计
      img_size_num = 0
      file_list = []
      pattern = re.compile('src="(\\{[^\\}]+\\})')
      for i, l in enumerate(qq_content_list): # l为每条记录
          # 根据图片大小总和按照$per大小 生成若干小文件
          if l.find('src="{') != -1:
              for v in pattern.findall(l): # v为dat名
                  if v in image_size_dict.keys():
                      img_size_num = img_size_num + image_size_dict[v]
                  if img_size_num > per * 1024:
                      # 记录每个文件开始的行号
                      file_list.append(i)
                      img_size_num = 0
      # 生成内容文件
      for k, v in enumerate(file_list):
          convert_name = base_dir +file_name_prefix + str(k + 1)
          html = ''
          # 第一个小文件
          if k == 0:
              for j in range(v+1):
                  html = html + qq_content_list[j] + split_str
              html = re.sub('src="(\\{[^"]+)"', src, html)
              new_content = header_str + html + html_end + split_str
          # 其他小文件
          else:
              for j in range(file_list[k - 1] + 1, v+1):
                  html = html + qq_content_list[j] + split_str
              html = re.sub('src="(\\{[^"]+)"', src, html)
              new_content = header2_str + html + html_end + split_str
          with open(convert_name +'.html', 'w', encoding="UTF-8") as f:
              f.write(new_content)
      # 最后一个小文件
      if v != len(qq_content_list):
          convert_name = base_dir + file_name_prefix + str(k + 2)
          html = ''
          for j in range(file_list[k] + 1, len(qq_content_list)):
              html = html + qq_content_list[j] + split_str
          html = re.sub('src="(\\{[^"]+)"', src, html)
          new_content = header2_str + html + html_end + split_str
          with open(convert_name +'.html', 'w', encoding="UTF-8") as f:
              f.write(new_content)
end = time.time()
print('不存在的图片：')
new_dat_list = list(dict.fromkeys(dat_list))
for i in new_dat_list:
  print(i)
print('{}个mht文件被转换，总耗时：{}s'.format(convert_num, end - start))

File: QQMhtSpliter.php
<?php
class QQMhtSpliter2 {
	const HTML_END_TAG = '</table></body></html>';
	private $per; //单位M，小于$per将不能分割

	private $dat_arr = []; //不存在的图片 保存在这里
	private $base_dir; //转换完成-目录
	private $convert_name = ''; //转换完成-文件名
	private $image_type = []; //每张图片的格式

	public function __construct($per = 40) {
		$this->per = $per;
	}

	public function convert() {
		$s = microtime(1);
		$m = memory_get_usage();
		$chat_file = $this->read_dir();
		$file_name_prefix = str_replace('.mht', '', $chat_file) . '_';
		$handle = fopen($chat_file, 'rb');
		for($i = 0; $i < 12; $i++) { 
			$row = fgets($handle);
			if(strpos($row, 'boundary="----=_NextPart') !== FALSE) {
				$sep = str_ireplace(['boundary=', '"'], '', trim($row));
			}
		}
		$html_end = FALSE;
		$txt_arr = [];
		$image_arr = [];
		$this->base_dir = './convert/';
		if (!is_dir($this->base_dir . 'images/')) {
			mkdir($this->base_dir . 'images/', 0777, TRUE);
		}
		if($this->read_file($handle)) {
			foreach($this->read_file($handle) as $key=>$l) {
				if(!$html_end && strpos($l, self::HTML_END_TAG) !== FALSE) {
					$html_end = TRUE;
					continue;
				}
				if($html_end) {
					if(strpos($l, 'Content-Type') !== FALSE) {
						$img_type = self::get_img_type($l);
						continue;
					}
					if(strpos($l, 'Content-Transfer-Encoding') !== FALSE) {
						continue;
					}
					if(strpos($l, 'Content-Location') !== FALSE) {
						preg_match("/(\{[^\}]+\}\.dat)/i", $l, $matches);
						$img_name = str_replace('.dat', '', $matches[1]);
						continue;
					}
					if(strpos($l, $sep) !== FALSE) {
						if(isset($n) && !empty($image_arr[$n])) {
							$base64 = implode('', $image_arr[$n]);
							$image_size[$img_name] = round(strlen($base64) / 1024, 2); //每张图片大小 单位kb
							$this->image_type[$img_name] = $img_type;
							$this->base64toimage($base64, $img_name, $img_type);
							unset($image_arr[$n]); //及时释放内存
						}
						$n = $key;
					} else {
						$image_arr[$n][] = $l;
					}
				} else {
					$txt_arr[] = $l;
				}
			}
		}
		fclose($handle);
		//$txt_arr第一个元素包含头部信息
		preg_match('/\d{4}-\d{2}-\d{2}<\/td><\/tr>/', $txt_arr[0], $b);
		$len = strpos($txt_arr[0], $b[0]) + strlen($b[0]);
		//头文件1
		$html_header = substr($txt_arr[0], 0, $len) . PHP_EOL;
		//头文件2
		$html_header_other = str_replace('height:24px;line-height:24px;padding-left:10px;margin-bottom:5px;', 'padding-left:10px;', $html_header);
		$html_header_other = str_replace('<tr><td><div style=padding-left:10px;>&nbsp;</div></td></tr>', '', $html_header_other);
		$html_header_other = preg_replace('/日期: \d{4}-\d{2}-\d{2}/', '&nbsp;', $html_header_other);
		$txt_arr[0] = substr($txt_arr[0], $len, strlen($txt_arr[0]));
		$qq_text = implode('', $txt_arr);
		if(filesize($chat_file) / 1024 / 1024 < $this->per) {
			$this->convert_name = str_replace('.mht', '', $chat_file);
			$qq_text = $this->src($qq_text);
			$qq_text = $html_header . $qq_text . self::HTML_END_TAG;
			file_put_contents($this->base_dir . $this->convert_name . '.html', $qq_text);
		} else {
			$qq_text_arr = explode(PHP_EOL, $qq_text);
			$image_size_sum = 0; //计算图片总大小
			$lines = []; //分隔完的文件 最后一条记录在$qq_text_arr中的所在行数
			foreach ($qq_text_arr as $l => $val) {
				//根据图片大小总和按照per大小 生成若干小文件
				if (strpos($val, 'src="{') !== FALSE) {
					preg_match_all('/src="(\{[^\}]+\})/i', $val, $matches);
					foreach ($matches[1] as $v) {
						if (isset($image_size[$v])) {
							$image_size_sum += $image_size[$v];
						}
					}
					if ($image_size_sum > $this->per * 1024) { //per生成每个文件的大小  第107行开始处理最后一个小文件(最后一个小文件有可能不够per * 1024)
						$lines[] = $l; //记录每个文件最后一行记录的行号
						$image_size_sum = 0;
					}
				}
			}
			//遍历分隔的文件，生成html文件
			foreach ($lines as $key => $line) {
				$html = '';
				$this->convert_name = $file_name_prefix . ($key + 1); //文件名从1开始命名
				//第一个小文件（头部和其他分割的头部有一些区别）
				if ($key == 0) {
					for ($i = 0; $i <= $line; $i++) {
						$html .= $qq_text_arr[$i] . PHP_EOL;
					}
					$html = $this->src($html);
					$new_content = $html_header . $html . self::HTML_END_TAG . PHP_EOL;
					//其余小文件
				} else {
					for ($i = $lines[$key - 1] + 1; $i <= $line; $i++) { //$lines[$key - 1]为上一个文件的最后一条所在行数，+1为当前文件第一条所在行数。 循环该文件第一条到最后一条记录
						$html .= $qq_text_arr[$i] . PHP_EOL;
					}
					$html = $this->src($html);
					$new_content = $html_header_other . $html . self::HTML_END_TAG . PHP_EOL;
				}
				file_put_contents($this->base_dir . $this->convert_name . '.html', $new_content);
			}
			//最后一个小文件（存在分割到最后，不够per * 1024的大小，总行数和分割之后最后一个文件的最后一条记录的行数对比）
			if ($line != count($qq_text_arr)) {
				$html = '';
				$this->convert_name = $file_name_prefix . ($key + 2); //这里的key 延用上面87行的key
				for ($i = $lines[$key] + 1; $i <= count($qq_text_arr) - 1; $i++) { //$lines[$key]为最后一个分割的文件最后一条所在行数，+1为当前文件第一条所在行数。 循环该文件第一条到最后一条记录
					$html .= $qq_text_arr[$i] . PHP_EOL;
				}
				$html = $this->src($html);
				$new_content = $html_header_other . $html . self::HTML_END_TAG . PHP_EOL;
				file_put_contents($this->base_dir . $this->convert_name . '.html', $new_content);
			}
		}
		//有些不存在的图片可能在多个分割文件都有出现，这里只显示1条，要显示全部屏蔽这行即可
		if($this->dat_arr) {
			echo PHP_EOL . "不存在的图片：" . PHP_EOL;
			foreach ($this->dat_arr as $dir=>$dat) {
				foreach($dat as $k=>$val) {
					echo $val . ' => ' . $k . PHP_EOL; //去重。 给出提示，在哪个分割的文件内（根据提示，可以手动去pc端或者手机端 定位到不显示的图片或表情，然后将其保存到images文件夹下，文件名用$val+'.gif'命名，包含'{','}'）
				}
			}
		}
		echo PHP_EOL . 'mht文件转换完成，总耗时:' . round(microtime(1) - $s, 3) . "s" . PHP_EOL;
		$unit = ['b', 'kb', 'mb', 'gb'];
		$size = memory_get_usage() - $m;
		echo PHP_EOL . '占用内存 '.round($size / pow(1024, ($i = floor(log($size, 1024)))), 2).' '.$unit[$i]. PHP_EOL;
	}

	//获取图片格式
	static private function get_img_type($image_type) {
		if (strpos($image_type, 'gif') !== FALSE) {
			$ext = '.gif';
		}
		if (strpos($image_type, 'png') !== FALSE) {
			$ext = '.png';
		}
		if (strpos($image_type, 'jpeg') !== FALSE || strpos($image_type, 'jpg') !== FALSE) {
			$ext = '.jpg';
		}
		return $ext;
	}

	private function base64toimage($base64, $k, $img_type) {
		file_put_contents($this->base_dir . 'images/' . $k . $img_type, base64_decode($base64));
	}

	private function src($html) {
		return preg_replace_callback('/src="(\{[^"]+)"/i', function($m) {
			$s = str_replace('.dat', '', $m[1]);
			if (isset($this->image_type[$s])) {
				return 'src="./images/' . $s . $this->image_type[$s] . '"';
			} else { //不存在的图片转成gif
				$this->dat_arr[$this->base_dir][$s] = str_replace('./\\', '', $this->base_dir . $this->convert_name) . '.html';
				return 'src="./images/' . $s . '.gif"';
			}
		}, $html);
	}

	private function read_file($handle) {
		while($line = fgets($handle)) {
			yield $line;
		}
	}

	private function read_dir() {
		$files = [];
		foreach(glob('./*') as $file) {
			if(strpos($file, '.mht') !== FALSE)
				$files[] = $file;
		}
		return str_replace('./', '', current($files));
	}
}

$opt = getopt("s:");
if(isset($opt['s']) && intval($opt['s'])) {
	$Spliter = new QQMhtSpliter2(intval($opt['s']));
} else {
	$Spliter = new QQMhtSpliter2();
}
$Spliter->convert();
